app/mbm/static/js/userlocations.js

```diff
@@ -0,0 +1,134 @@
+        // Load user locations from localstorage
+        this.locations = getUserLocations() || {}
+
+        // Render user locations
+        for (const [name, location] of Object.entries(this.locations)) {
+            this.renderSavedLocation(name, location)
+        }
+    }
+
+    createLocationPopup() {
+        const popup = L.popup()
+        const $form = $(`
+            <form>
+                <h4>Add new saved location?</h4>
+                <label for="locationName">Name</label>
+            </form>`
+        )
+        const $nameInput = $('<input id="locationName" name="locationName">')
+        this.$nameInput = $nameInput
+        const $saveButton = $('<button class="btn btn-primary btn-block">Save location</button>')
+        $form.append($nameInput).append($saveButton)
+        $form.submit(this.handleFormSubmission.bind(this))
+
+        this.form = $form[0]
+        popup.setContent(this.form)
+        return popup
+    }
+
+    handleFormSubmission(event) {
+        event.preventDefault()
+
+        const name = this.$nameInput.val()
+        const latlng = this.addLocationPopup.getLatLng()
+
+        this.addLocation(name, latlng.lat, latlng.lng)
+        this.form.reset()
+        this.closeAddLocationPopup()
+    }
+
+    // Create a popup with a form for users to save a new location to the map
+    showAddLocationPopup(clickEvent) {
+        this.addLocationPopup
+            .setLatLng(clickEvent.latlng)
+            .openOn(this.map)
+    }
+
+    // Close the new location popup
+    closeAddLocationPopup() {
+        this.map.closePopup(this.addLocationPopup)
+    }
+
+    // Add a new location to the map and update the locations in localstorage
+    addLocation(name, lat, lng) {
+        const location = { lat, lng }
+        this.locations[name] = location
+        this.renderSavedLocation(name, location)
+        this.saveLocations()
+    }
+
+    // Save user locations to local storage
+    saveLocations() {
+        saveUserLocations(this.locations)
+    }
+
+    // "render" a location to the map by creating a marker at that location,
+    // with a popup containing controls to navigate to/from that marker or to
+    // delete it
+    renderSavedLocation(name, location) {
+        const marker = L.marker([location.lat, location.lng], {
+            title: name,
+            // Make sure these markers sit above the regular source/target markers
+            zIndexOffset: 100,
+        })
+
+        const popup = L.popup()
+        const sourceButton = $(`<button class="btn btn-primary btn-block">Directions from ${name}</button>`).click(() => {
+            this.app.setSourceLocation(location.lat, location.lng, name)
+            this.map.closePopup(popup)
+        })
+        const targetButton = $(`<button class="btn btn-primary btn-block">Directions to ${name}</button>`).click(() => {
+            this.app.setTargetLocation(location.lat, location.lng, name)
+            this.map.closePopup(popup)
+        })
```

Comment:
**[Nitpick, non-blocking]** From a user perspective, I almost wonder if we want to automatically run the search when the user hits one of these buttons, provided that the other option is already set (e.g. if the user clicks "Directions to home" while the `source` field is populated, we would automatically run the search)? Not a blocking change by any means, but might be a nice UX improvement if you agree!

Reply:
That sounds like a great improvement. I'll open an issue for it!

Reply:
https://github.com/jeancochrane/mellow-bike-map/issues/44